[ML] Implement MSLE loss function for regression

valeriy42 opened the pull request:
This PR implements mean squared logarithmic error loss function for regression. It also adds `CLoss::isRegression()` function to distinguish between regression and classification loss functions.

Relates to #973

tveasey commented on lib/maths/CBoostedTreeLoss.cc:
```diff
@@ -366,6 +366,144 @@ CArgMinMultinomialLogisticLossImpl::objectiveGradient() const {
+                minLogWeight = std::min(minLogWeight,
+                                        CBasicStatistics::mean(bucketActual)(2));
+                maxLogWeight = std::max(maxLogWeight,
+                                        CBasicStatistics::mean(bucketActual)(2));
```
I wonder if it might be worthwhile naming constants, i.e. `static constexpr std::size_t PREDICTION = 0`, ..., `static constexpr std::size_t ERROR = 2`. When I first read this I thought it was an error, but of course you're actually reading the error out of the bucket, so doing the right thing if this was it would make this clearer I think
```suggestion
                                        CBasicStatistics::mean(bucketActual)(ERROR));
```

valeriy42 replied:
Fixed

tveasey commented on include/maths/CBoostedTreeLoss.h:
```diff
@@ -384,6 +445,43 @@ class MATHS_EXPORT CMultinomialLogisticLoss final : public CLoss {
+//!   \f$\displaystyle l_i(p) = c_i + w_i(p - a_i)^2\f$
+//! </pre>
+//! where \f$w_i = \frac{\log(1+p_i) - \log(1+a_i)}{(1+p_i)(p_i-a_i)}\f$ and \f$c_i\f$
+//! is choosen so \f$l_i(p_i) = (\log(1+p_i) - \log(1+a_i))^2\f$.
```
```suggestion
//! is chosen so \f$l_i(p_i) = (\log(1+p_i) - \log(1+a_i))^2\f$.
```

valeriy42 replied:
Fixed

tveasey commented:
retest

tveasey commented:
One other comment, this shouldn't be a non-issue and should appear in the change log.

valeriy42 commented on include/maths/CBoostedTreeLoss.h:
```diff
@@ -66,6 +67,66 @@ class MATHS_EXPORT CArgMinMseImpl final : public CArgMinLossImpl {
 };
 
+//! \brief Finds the value to add to a set of predictions which approximately
+//! minimises the regularised Mean squared logarithmic error (MSLE).
```
Fixed

valeriy42 commented on include/maths/CBoostedTreeLoss.h:
```diff
@@ -66,6 +67,66 @@ class MATHS_EXPORT CArgMinMseImpl final : public CArgMinLossImpl {
+        double predictionBucketWidth = m_ExpPredictionMinMax.range() /
+                                       static_cast<double>(m_Buckets.size());
+        double actualBucketWidth = m_LogActualMinMax.range() /
+                                   static_cast<double>(m_Buckets[0].size());
```
Fixed

valeriy42 commented on include/test/CDataFrameAnalyzerTrainingFactory.h:
```diff
@@ -125,7 +142,12 @@ class TEST_EXPORT CDataFrameAnalyzerTrainingFactory {
+                    break;
                 case E_BinaryClassification:
+                    appendPrediction(*frame, weights.size(), prediction, expectedPredictions);
+                    break;
```
You are right, it was important before but seems to be superfluous now. I remove the switch case.

valeriy42 commented on lib/api/unittest/CDataFrameAnalyzerTrainingTest.cc:
```diff
@@ -570,6 +570,67 @@ BOOST_AUTO_TEST_CASE(testRunBoostedTreeRegressionTrainingWithStateRecovery) {
 
+BOOST_AUTO_TEST_CASE(testRunBoostedTreeRegressionTrainingMsle) {
+
+    // Test that MSLE on scaled values predicts correct values same as MSE in testRunBoostedTreeRegressionTraining.
```
Fixed.

valeriy42 commented on lib/maths/CBoostedTreeLoss.cc:
```diff
@@ -366,6 +366,144 @@ CArgMinMultinomialLogisticLossImpl::objectiveGradient() const {
+}
+
+CArgMinMsleImpl::TDoubleVector CArgMinMsleImpl::value() const {
+    std::function<double(double)> objective;
```
Fixed

valeriy42 commented on lib/maths/CBoostedTreeLoss.cc:
```diff
@@ -366,6 +366,144 @@ CArgMinMultinomialLogisticLossImpl::objectiveGradient() const {
 }
+
+CArgMinMsleImpl::CArgMinMsleImpl(double lambda)
+    : CArgMinLossImpl{lambda}, m_Buckets(128) {
```
If I reduce the bucket size, I would need to relax the test containts since `CBoostedTreeLossTest/testMsleArgminObjective` starts to fail then.

tveasey replied:
I think this is ok. For example, I'm a bit worried about the memory this will use for a large tree. We don't actually monitor this memory at present (something we should perhaps do in a follow on PR).

valeriy42 replied:
I changed the bucket size to 32 and relaxed the unit test limits.

valeriy42 commented:
Thank you for your review @tveasey I addressed all your comments and added handling negative target values. Please, have another pass when you have time.

valeriy42 commented:
retest
